service_configuration_lib/cached_view.py

```diff
@@ -0,0 +1,223 @@
+
+import pyinotify
+
+DEFAULT_SOA_DIR = '/nail/etc/services'
+DEFAULT_PROCESS_EVENTS_LIMIT = 100
+
+log = logging.getLogger(__name__)
+
+
+class BaseCachedView(ABC):
+    """This is the interface for ConfigsFileWatcher.
+    Implementation of that interfaces should implemente 2 hooks on creation/removal of files.
+    """
+
+    @abstractmethod
+    def add(self, path: str, service_name: str, config_name: str, config_suffix: str) -> None:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def remove(self, path: str, service_name: str, config_name: str, config_suffix: str) -> None:
+        raise NotImplementedError()
+
+
+_ServiceConfig = namedtuple('_ServiceConfig', ['service_name', 'config_name', 'config_suffix'])
+
+
+class ConfigsFileWatcher:
+    """This is a cache of the smartstack.yaml config file contents found in
+    /nail/etc/services, with keys of the form '<servicename>.<namespace>'.
+
+    What's special about this cache is that we watch for inotify events on these
+    files, processing these events whenever the `process_events` method is
+    invoked.
+
+    The benefit of the explicit event processing is that it obviates the need
+    for a background thread (which would bring its own set of problems).  The
+    downside is that the `process_events` method must be periodically invoked
+    in order to avoid overflowing the event queue.
+
+    How frequently should we invoked the `process_events` method?  I've measured
+    10 to 20 events per minute, and a queue size (max_queued_events) of 16384.
+    This gives us about half a day of non-processing before the queue overflows
+    and the cache becomes inconsistent.  In the unlikely case that the queue
+    does overflow, then the cache will resetupt itself.
+
+    Args:
+    :param configs_view: implementation of BaseCachedView interface
+    :param configs_folder: Optional, path to configs root folder, `/nail/etc/services`
+    :param services_names: Optional, list of service names to watch, can be wildcard or exact name
+    :param configs_names: Optional, list of config names to watch
+    :param configs_suffixes: Optional, list of file extentions to watch
+    :param exclude_folders_filters: Optional, filter of masks to exclude folders from watching
+    """
+
+    def __init__(
+        self,
+        configs_view: BaseCachedView,
+        configs_folder: str = DEFAULT_SOA_DIR,
+        services_names: Sequence[str] = ['*'],
+        configs_names: Optional[Sequence[str]] = None,
+        configs_suffixes: Optional[Sequence[str]] = None,
+        exclude_folders_filters: Sequence[str] = ['*.~tmp~'],
+    ) -> None:
+        super().__init__()
+        self.configs_view = configs_view
+        self._configs_folder = configs_folder
+        self._services_names = services_names
+        self._configs_names = None if configs_names is None else set(configs_names)
+        self._configs_suffixes = None if configs_suffixes is None else set(configs_suffixes)
```

Comment:
Yes, if it's None - library won't filter anything. If that's will be empty set/list, which actually makes not much sense it will filter everything, because nothing will match.